noise/README.md

```diff
@@ -0,0 +1,400 @@
+messages. The payload includes the peer's public libp2p identity key, and a
+signature over the static Noise public key, produced with the identity key. This
+simple certificate proves that the possessor of the Noise static keypair is also
+in possession of the private libp2p identity key.
+
+The lifetime of the Noise static keypair is not specified. Implementations MAY
+allow noise-libp2p to be initialized with a stored static key to extend the key
+lifetime beyond a single invocation of a given application, in which case it is
+the application's responsibility to store the key. If no static key is provided
+at initialization, noise-libp2p will generate a new X25519 keypair and use it
+for the lifetime of the process.
+
+### The noise-libp2p Handshake Payload
+
+As mentioned, noise-libp2p uses a static keypair that is distinct from the
+"identity keypair" used to derive a peer's peer id. As a result, we must
+authenticate Noise static keys in a way that proves they belong to the correct
+libp2p peer. To do so, noise-libp2p includes a simple "certificate" in the
+payload of Noise handshake messages.
+
+The noise-libp2p handshake payload is a [protobuf][protobuf] message with the
+following schema:
+
+``` protobuf
+message Identity {
+	bytes pubkey = 1;
+	bytes signature = 2;
+}
+
+```
+
+The `pubkey` field contains the public libp2p identity key, encoded as described
+in the [peer id spec][peer-id-spec].
+
+The `signature` field contains a signature of the Noise static public key,
+produced by the private libp2p identity key. The Noise static key is encoded
+into a byte array according to the rules defined in [section 5 of RFC
+7748][rfc-7748-sec-5] and signed as described in the [peer id
+spec][peer-id-spec].
+
+Each peer MUST send the noise-libp2p handshake payload in the first encrypted
+Noise handshake message that they send. Encrypting the handshake payload is
+required to avoid replay attacks, as there is no timestamp or other validity
+criteria in the payload itself apart from the signature.
+
+Noise handshake payloads are encrypted if they occur after at least one DH
+exchange. For the `XX` handshake, the first handshake message is sent in
+plaintext with no payload, and encrypted payloads are sent with the following
+two messages.
+
+Below are the supported handshake patterns annotated with the payload that is
+sent during each message (if any).
+
+Because the initiator and responder roles may change during the `IK` /
+`XXfallback` pattern, [Alice and Bob][npf-alice-and-bob] terminology is used. In
+all cases, Alice initiates the first handshake attempt (either `XX` or `IK`).
+Bob MAY initiate an `XXfallback` handshake following a failed `IK` attempt. The
+`XXfallback` handshake is written in "Bob-initiated form," with the direction of
+the arrows reversed from the canonical (Alice-initiated) notation.
+
+```
+XX:
+  -> e
+  <- e, ee, s, es   [Alice's identity payload]
+  -> s, se          [Bob's identity payload]
```

Comment:
Are Alice and Bob the wrong way around here?

Reply:
thank you, yes :)